src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,191 @@
+  todo: Todo,
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const { onNewError, setErrorMessage } = useContext(ErrorContext);
+  const {
+    todos,
+    setTodos,
+    todoIdsWithLoader,
+    setTodoIdsWithLoader,
+  } = useContext(TodosContext);
+  const { title, completed, id } = todo;
+  const [isBeingEdited, setIsBeingEdited] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+  const editedInputRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    if (editedInputRef.current) {
+      editedInputRef.current.focus();
+    }
+  }, [isBeingEdited]);
+
+  const handleTodoDelete = () => {
+    setTodoIdsWithLoader(prevTodoIds => [...prevTodoIds, id]);
+    setErrorMessage(ErrorMessage.None);
+    deleteTodo(id)
+      .then(() => {
+        setTodos(prevTodos => prevTodos
+          .filter(({ id: todoId }) => id !== todoId));
+        setIsBeingEdited(false);
+      })
+      .catch(() => onNewError(ErrorMessage.UnableDelete))
+      .finally(() => setTodoIdsWithLoader(
+        prevTodoIds => prevTodoIds.filter((todoId) => todoId !== id),
+      ));
+  };
+
+  const handleChangeOfTitle = ((
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setNewTitle(event.target.value);
+  });
+
+  const handleNewTitleSubmit = (
+    event?: React.FormEvent<HTMLFormElement>
+    | React.FocusEvent<HTMLInputElement, Element>,
+  ) => {
+    event?.preventDefault();
+    const trimmedTitle = newTitle.trim();
+
+    if (!trimmedTitle) {
+      handleTodoDelete();
+
+      return;
+    }
+
+    if (trimmedTitle === title) {
+      setIsBeingEdited(false);
+
+      return;
+    }
+
+    setTodoIdsWithLoader(prevTodoIds => [...prevTodoIds, id]);
+    setErrorMessage(ErrorMessage.None);
+
+    updateTodo(id, { title: trimmedTitle })
+      .then(() => {
+        const todosCopy = [...todos];
+        const searchedTodo = todos.find(
+          ({ id: todoId }) => todoId === id,
+        ) as Todo;
+
+        searchedTodo.title = trimmedTitle;
+
+        setTodos(todosCopy);
+        setIsBeingEdited(false);
+      })
+      .catch(() => onNewError(ErrorMessage.UnableUpdate))
+      .finally(() => {
+        setTodoIdsWithLoader(
+          prevTodoIds => prevTodoIds.filter((todoId) => todoId !== id),
+        );
+      });
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setIsBeingEdited(false);
+      setNewTitle(title);
+    }
+  };
+
+  const handleTodoToggle = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTodoIdsWithLoader(prevTodoIds => [...prevTodoIds, id]);
+    setErrorMessage(ErrorMessage.None);
+    updateTodo(id, { completed: !completed })
+      .then(() => {
+        const todosCopy = [...todos];
+        const searchedTodo = todos.find(
+          ({ id: todoId }) => todoId === id,
+        ) as Todo;
+
+        searchedTodo.completed = !event.target.checked;
+
+        setTodos(todosCopy);
```

Comment:
```suggestion
        setTodos(prevTodos => (
          prevTodos.map(todo => todoId === todo.id ? updatedTodo : todo)
        ));
```
Use `map` instead of `find`.